### ISSUE TITLE
Better approach for CompactNumberFormat in QuantityFormat

### ISSUE THREAD

**Issue by keilw** [unitsofmeasurement/indriya#349]
After support and proper handling of the new `CompactNumberFormat` in Java 12 was added by #251 the idea was to control how that `NumberFormat` is created and made part of a `NumberDelimiterQuantityFormat`, but limitations of the Java spec (as discussed in #346) would not allow creating an instance via
`getCompactInstance(FormatBehavior, NumberFormat.Style)` which it ideally should because that method signature won't work before Java 12. That makes `getCompactInstance(FormatBehavior)` pretty useless because it is merely a convenience wrapper for one of at least two format style cases. 
And instead a call to either
```
NumberDelimiterQuantityFormat.Builder()
            .setNumberFormat(NumberFormat.getCompactNumberInstance(Locale.ROOT, NumberFormat.Style.SHORT))
            .setUnitFormat(SimpleUnitFormat.getInstance()).build();
```
or
```
NumberDelimiterQuantityFormat.Builder()
            .setNumberFormat(NumberFormat.getCompactNumberInstance(Locale.ROOT, NumberFormat.Style.LONG))
            .setUnitFormat(SimpleUnitFormat.getInstance()).build();
```
shold be made explicitly.

We likely need the special Java 12 version of `NumberDelimiterQuantityFormat` to handle CompactNumberFormat differently if required, but all the factory methods especially those specific to Java 12 and above could be eliminated that way.

**Comment by keilw** [comment]
@andi-huber, @desruisseaux What do you think about this and could either of you also help or are you too busy with other tasks?

**Comment by keilw** [comment]
Martin, No worries, I think this is small enough to even do it on my own if nobody else had enough resources. And it would solve the "signature concerns" (as tiny and insignificant as they actually are, because most people may not even know what `CompactNumberFormat` is or use it ;-) without introducing that "stub" method for lower Java versions.

**Comment by keilw** [comment]
Andi, no worries, then I'd look into it. I know you are more into "Arithmetics" and your contributions (currently the second-most in this repo) are highly appreciated, so no problem leaving that to others.
We learned especially from that other ticket #346 it's often better not to put one's oar in (German: "seinen Senf dazugeben") everywhere ;-)

**Comment by andi-huber** [comment]
Well quite frankly - regarding the elephant in the room (346) - I have a different perspective. I don't mind a technical discourse, where participants have an opportunity to bring their ideas and then let the better argument win. But to watch this tendency towards insults on a personal level is painful, exhausting and is something I don't want to be associated with. I am here for the fun of it and to learn a few things.

**Comment by keilw** [comment]
I did not insult him, and the guy who raised the ticket also wrote the kind of stuff quoted which may be you learn in books or interview questions by the likes of Google, but in reality even the JDK itself does not apply.
"foot-soldier" is what people supporting customers have to do in the "trenches" just like people in WWI...
I was a Developer Relationship "something" just like @nipafx is now. And even if our project may not pay a Million Euros or Pounds to Oracle we still deserve a polite answer (even if it would be "I don't know, but I could ask someone") and not this shabby "Go do yourself" he gave me. Only to have his boss or whatever the "Line of Duty" might be answer it after it may have been escalated to the right level of responsibility.
The real insults happen by bad actors who place their malware, ransomware etc. in the mailing lists, mostly at Eclipse (ours at units-users or units-dev may not seem attractive enough but I also got some strange requests on the JSR 354 mailing lists that might be from a front company for some fraud or cybercrime because after asking him about a missing link on their page or something I never heard from him since) and Jakarta EE. 
One almost could have thought @anthonyvdotbe to have similarly sinister goals, but it seems he's more harmless than those willfully placing malware or phishing links into what they mimick as Jakarta challenges, tutorials or quizzes.
He made a single contribution to Jakarta REST but all of it looks like fixing typos in Asciidoc, not an actual piece of code.
So a "tech writer" but not an actual coder, otherwise you would also see more like the 2k annual committs you got or I do and almost everyone else and not 59 ;-)
Ok Martin also does not push so regularly on GitHub, but he commits a lot to Apache SIS and I don't think the mirrors always reflect the original authors on GH, at least I cannot say for sure.

**Comment by keilw** [comment]
Btw @andi-huber how did the work with GeoTools go? You will recall they also had a bit of an unfriendly tone towards us and especially @desruisseaux even more than myself, and that was based on his description mostly because he and SIS "dared to create another implementation". 
So you sometimes get these kinds of things and if you spent anough time in the JCP EC like @dautelle and I did, you learn even more of those things and people like Rod Johnson calling the EC the "political commissars" like those in Communist countries only to be invited by Oracle or I think it was still Sun at the time into that very same EC in a ratified seat. He just showed up at one F2F for about 15 minutes or so. 10 out of that I had a conversation with him, then his phone rang and he left without ever being seen again. Leading to the "Lex Johnson" in the JCP about participation. Something we also consider in the Jakarta EE WG because especially the representatives of LJC were notoriously absent from Spec Committee calls (either because they wear "too many hats" including the JCP EC one or because their new bosses at Microsoft are at their neck with corporate projects?)

**Comment by desruisseaux** [comment]
I do not think that GeoTools peoples are unfriendly in general. The friction between myself and some other core members had a longer history that started before we decided to continue our developments outside. It is an history of different priorities, code review not properly done, commercial interests causing pressure, _etc._ Those things happen, it just get more intense when we invested a lot in a project.

On latest discussions, I think that authority arguments should be avoided because they create frustrations, because they usually come from a different context not necessarily applicable to the discussion. Instead if a technical argument got no answer or if the answer suggests that there is a misunderstanding, I suggest to divide the technical argument into simpler arguments until we reach a common understanding.

**Comment by keilw** [comment]
What is that whole "authority" thing? At least in the JCP EC or Jakarta Spec Committee I get elected to a higher "authority" and that I pointed out myself and others were confronted with say JPMS or every Java SE Spec since Java 7.
I cannot undo that and I had to witness and even help moderate (with my EC "hat") sometimes much more hostile and foul-mouthed arguments between e.g. the OSGi people and OpenJDK/JPMS team or between JSRs like `@Inject` (especially Bob Lee who got the name "Crazy Bob" for a reason) and the first version of CDI.

**Comment by desruisseaux** [comment]
The word "authority" was not refering to anyone in particular. "Authority argument" is a generic name for a particular kind of argument. When a discussion uses the opinion of an authority on a topic as evidence to support an argument, this is called "authority argument". The authority is not necessarily any of the peoples participating directly to the discussion.

**Comment by keilw** [comment]
Ok but either way I or @dautelle cannot change that we had "multiple hats" including those on the other side of almost every spec including the Java SE ones, thus having seen and reviewed those specs long before others like @nipafx were involved was naturally part of our duties. 
And as mentioned here before, I won't remind them explicitly because they have many other things to do (especially on the Jakarta EE side) but if the JCP EC had the same attitude and eagerness it had around 10 years ago, THEY would have told us that a class likely violated the spec because they keep reviewing and approving all those specs on a regular basis, but many of them do not care as much about the technical aspect as others including myself did. I pretty much made all my votes based on technical arguments even if I was the only one abstaining or voting "No" sometimes, other cases like that JPMS vote saw a majority against it, similar to what we had with JSR 275 although for a combination of different reasons as you may remember.
I don't see @nipafx unblocked yet but I guess those 3 days might be over soon.

**Comment by keilw** [comment]
I also created the page for UCAR and JSR 108: http://unitsofmeasurement.github.io/2021/20yearsjavaxunits.html
A little late, but it also anticipates a possible implementation by NetCDF. @desruisseaux Is there any development there or prospect it might happen?
Unless the reasonably constructive discussion about toolchains for MP-JARs also used in Glassfish https://github.com/eclipse-ee4j/jakartaee-platform/issues/329#issuecomment-840524012 brings a much better alternative to the toolchain plugin used here, I don't see how we could or should change that (as #348 suggested) but for any implementation including Indriya itself, Seshat (2.x) or maybe a future NetCDF one we may well add signature checks or other verifications to the TCK.

**Comment by desruisseaux** [comment]
I'm not aware of work regarding `javax.measure` implementation by UCAR. I think this is a "nice to have" feature for them, but with no short term plan yet. They seem busy with a major upgrade of the netCDF library for now.

**Comment by keilw** [comment]
Never mind, if they get there nice, otherwise no problem either. Closing this, because all the methods are removed in favor of the described approach in [java12/QuantityFormatDemo](https://github.com/unitsofmeasurement/uom-demos/blob/master/console/java12/src/main/java/tech/uom/demo/java12/format/QuantityFormatDemo.java). As #301 explained the JDK bug we helped workaround when fixing it, the separate demos are also not required anymore.
